### PR TITLE
Verify charmhub subordinates

### PR DESCRIPTION
**Pull request by SimonRichardson**
<strike>Requires https://github.com/juju/juju/pull/12776 to land first, will rebase once it's landed.</strike>

Upgrade series should work with subordinates. The work here was to
verify that and remove duplicated code.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy ubuntu --series=bionic
$ juju deploy nrpe
$ juju add-relation ubuntu nrpe
$ juju upgrade-series 0 prepare focal
WARNING: This command will mark machine "0" as being upgraded to series "focal".
This operation cannot be reverted or canceled once started.
Units running on the machine will also be upgraded. These units include:
  ubuntu/0

Leadership for the following applications will be pinned and not
subject to change until the "complete" command is run:
  ubuntu

Continue [y/N]?y
machine-0 validation of upgrade series from "bionic" to "focal"
machine-0 started upgrade series from "bionic" to "focal"
ubuntu/0 pre-series-upgrade hook running
ubuntu/0 pre-series-upgrade completed
nrpe/0 pre-series-upgrade hook running
nrpe/0 pre-series-upgrade hook not found, skipping
machine-0 binaries and service files written

Juju is now ready for the series to be updated.
Perform any manual steps required along with "do-release-upgrade".
When ready, run the following to complete the upgrade series process:

juju upgrade-series 0 complete
```

### CODE REVIEW

**Review by manadart — CHANGES_REQUESTED** [review]
I think this needs some reorganisation.

We have a method on the API, `validateMachine`, which does not call through to validator methods.

Then we have `validateApplication`, which calls the validator `ValidateSeries` (which is machine-centric) and `ValidateApplications`.

We should make validation wholly the concern of the validator and call it's methods from the facade methods without implementing the private methods that share functionality between them and the validator.

**Comment by SimonRichardson** [comment]
@manadart fair comment about `ValidateMachine` is valid, I'll move that out into a better location

File: apiserver/facades/client/machinemanager/upgrade_series.go [hunk 1]
```diff
@@ -324,8 +329,46 @@ func makeUpgradeSeriesValidator(client CharmhubClient) upgradeSeriesValidator {
 
 // ValidateSeries validates a given requested series against the current
 // machine series.
-func (s upgradeSeriesValidator) ValidateSeries(requested, machine, tag string) error {
-	return validateSeries(requested, machine, tag)
+func (s upgradeSeriesValidator) ValidateSeries(requestedSeries, machineSeries, machineTag string) error {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/client/machinemanager/upgrade_series.go:332]: I want to remove `machineTag` as it's used to augment an error, which should be done with `errors.Annotatef`. For now I'll leave this as is.

**Comment by manadart** [comment]
QA passed, but shouldn't the review information include all units?
```
Units running on the machine will also be upgraded. These units include:
  ubuntu/0
```
Shouldn't it read like this?
```
Units running on the machine will also be upgraded. These units include:
  ubuntu/0
  nrpe/0
```

**Comment by SimonRichardson** [comment]
> QA passed, but shouldn't the review information include all units?
> 
> ```
> Units running on the machine will also be upgraded. These units include:
>   ubuntu/0
> ```
> 
> Shouldn't it read like this?
> 
> ```
> Units running on the machine will also be upgraded. These units include:
>   ubuntu/0
>   nrpe/0
> ```

This is solved with the last commit:

```sh
$ juju upgrade-series 0 prepare focal
WARNING: This command will mark machine "0" as being upgraded to series "focal".
This operation cannot be reverted or canceled once started.
Units running on the machine will also be upgraded. These units include:
  ubuntu/0
  nrpe/0

Leadership for the following applications will be pinned and not
subject to change until the "complete" command is run:
  nrpe
  ubuntu

Continue [y/N]?
```

**Review by manadart — APPROVED** [review]
Nice.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**